code/notificationbuilder/src/main/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/ProductCatalogNotificationBuilder.kt

```diff
@@ -0,0 +1,335 @@
+        // downloads all the catalog item images then sets the currently selected product
+        // within the main product catalog image view
+        val downloadedImageCount = populateCenterImage(
+            context,
+            trackerActivityClass,
+            expandedLayout,
+            pushTemplate,
+            catalogItems
+        )
+        if (downloadedImageCount != catalogItems.size) {
+            Log.error(
+                LOG_TAG,
+                SELF_TAG,
+                "Failed to download all images for the product catalog notification."
+            )
+            throw NotificationConstructionFailedException("Failed to download all images for the product catalog notification.")
+        }
+
+        // set the product title, description, and price
+        expandedLayout.setTextViewText(
+            R.id.product_title,
+            catalogItems[pushTemplate.currentIndex].title
+        )
+        expandedLayout.setTextViewText(
+            R.id.product_description,
+            catalogItems[pushTemplate.currentIndex].body
+        )
+        expandedLayout.setTextViewText(
+            R.id.product_price,
+            catalogItems[pushTemplate.currentIndex].price
+        )
+
+        // setup the CTA button
+        setupCtaButton(context, trackerActivityClass, expandedLayout, pushTemplate)
+
+        // populate product thumbnails
+        populateThumbnails(
+            context,
+            broadcastReceiverClass,
+            channelIdToUse,
+            expandedLayout,
+            catalogItems,
+            pushTemplate
+        )
+
+        return notificationBuilder
+    }
+
+    /**
+     * Sets the product catalog center image and its click action.
+     *
+     * @param context the application [Context]
+     * @param trackerActivityClass the [Class] of the activity to set in the created pending intent for tracking purposes
+     * @param expandedLayout the [RemoteViews] object containing the expanded layout of the push template notification
+     * @param pushTemplate the [ProductCatalogPushTemplate] object containing the product catalog push template data
+     * @param catalogItems the list of [ProductCatalogPushTemplate.CatalogItem] objects containing the product catalog items
+     * @return the number of images downloaded
+     */
+    private fun populateCenterImage(
+        context: Context,
+        trackerActivityClass: Class<out Activity>?,
+        expandedLayout: RemoteViews,
+        pushTemplate: ProductCatalogPushTemplate,
+        catalogItems: List<ProductCatalogPushTemplate.CatalogItem>
+    ): Int {
+        Log.trace(
+            LOG_TAG,
+            SELF_TAG,
+            "Populating center image for product catalog notification."
+        )
+
+        val catalogImageUris = catalogItems.map { it.img }
+        val downloadedImageCount = PushTemplateImageUtils.cacheImages(catalogImageUris)
+        if (downloadedImageCount == catalogImageUris.size) {
```

Comment:
How do you feel about moving the logic in lines 151-153 to the constructor and throwing an exception if all catalogItems images are not downloaded? Subsequently we call `populateCenterImage` knowing images exist in cache

Reply:
yup, can update this